marker/processors/high_quality_text.py

```diff
@@ -0,0 +1,207 @@
+ ]
+}
+```
+
+Output:
+
+```json
+{
+ "corrected_lines": [
+  "Adversarial training (AT) [23], which aims to minimize\n",
+  "the model's risk under the worst-case perturbations, is cur-\n",
+  "rently the most effective approach for improving the robust-\n",
+  "ness of deep neural networks. For a given neural network\n",
+  "$f(x, w)$ with parameters $w$, the optimization objective of\n",
+  "AT can be formulated as follows:\n"
+ ]
+}
+```
+
+**Input:**
+
+"""
+
+    def __init__(self, config=None):
+        super().__init__(config)
+
+        self.model = None
+        if not self.high_quality:
+            return
+
+        if self.google_api_key is None:
+            raise ValueError("Google API key is not set")
+
+        genai.configure(api_key=self.google_api_key)
+        self.model = genai.GenerativeModel(self.model_name)
+
+    def __call__(self, document: Document):
+        if not self.high_quality or self.model is None:
+            return
+
+        self.rewrite_blocks(document)
+
+    def rewrite_blocks(self, document: Document):
+        pbar = tqdm(desc="High quality text processor")
+        with ThreadPoolExecutor() as executor:
+            for future in as_completed([
+                executor.submit(self.process_block_rewriting, document, page, block)
+                for page in document.pages
+                for block in page.contained_blocks(document, (BlockTypes.TextInlineMath, BlockTypes.Handwriting))
+            ]):
+                future.result()  # Raise exceptions if any occurred
+                pbar.update(1)
+
+        pbar.close()
+
+    def process_block_rewriting(self, document: Document, page: PageGroup, block: Block):
+        SpanClass: Span = get_block_class(BlockTypes.Span)
+
+        text_lines = block.contained_blocks(document, (BlockTypes.Line,))
+        extracted_lines = [line.formatted_text(document) for line in text_lines]
+
+        prompt = self.gemini_rewriting_prompt + '```json\n`' + json.dumps({"extracted_lines": extracted_lines}, indent=2) + '`\n```\n'
+        image = self.extract_image(page, block)
+        response_schema = content.Schema(
+            type=content.Type.OBJECT,
+            enum=[],
+            required=["corrected_lines"],
+            properties={
+                "corrected_lines": content.Schema(
+                    type=content.Type.ARRAY,
+                    items=content.Schema(
+                        type=content.Type.STRING,
+                    ),
+                )
+            },
+        )
+
+        response = self.generate(prompt, image, response_schema)
+        corrected_lines = []
+        if response and "corrected_lines" in response:
+            corrected_lines = response["corrected_lines"]
+
+        if corrected_lines and len(corrected_lines) == len(extracted_lines):
+            for text_line, corrected_text in zip(text_lines, corrected_lines):
+                span_block = page.add_full_block(
+                    SpanClass(
```

Comment:
This will include markdown formatting inside the text, right?  Like `*` and `**`.  But we store things in html - so this will cause some issues with html and json output, right?

Reply:
Oh right, we should switch this to HTML

Reply:
fixed it! thank you!